secrets/azurekeyvault/akv.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+)
+
+var (
+	// KeyVault URI suffix
+	keyVaultEndpointSuffix = "vault.azure.net"
+	// Map of HTTP Status Code to go-cloud ErrorCode
+	errorCodeMap = map[int]gcerrors.ErrorCode{
+		200: gcerrors.OK,
+		400: gcerrors.InvalidArgument,
+		401: gcerrors.PermissionDenied,
+		404: gcerrors.NotFound,
+		408: gcerrors.DeadlineExceeded,
+		429: gcerrors.ResourceExhausted,
+		500: gcerrors.Internal,
+		501: gcerrors.Unimplemented,
+	}
+)
+
+type (
+	keeper struct {
+		client       *keyvault.BaseClient
+		keyVaultName string
+		keyName      string
+		keyVersion   string
+		options      *KeeperOptions
+	}
+
+	// KeeperOptions provides configuration options for encryption/decryption operations.
+	KeeperOptions struct {
+		Algorithm string
+	}
+)
+
+// Dial gets a new *keyvault.BaseClient, see https://godoc.org/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#BaseClient
+func Dial() (*keyvault.BaseClient, error) {
+	auth, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return nil, err
+	}
+
+	client := keyvault.NewWithoutDefaults()
+	client.Authorizer = auth
+	client.Sender = autorest.NewClientWithUserAgent(useragent.AzureUserAgentPrefix("secrets"))
+	return &client, nil
+}
+
+// NewKeeper returns a *secrets.Keeper that uses Azure keyVault.
+// List of Parameters:
+//	- client: *keyvault.BaseClient instance, see https://godoc.org/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#BaseClient
+//  - keyVaultName: string representing the KeyVault name, see https://docs.microsoft.com/en-us/azure/key-vault/common-parameters-and-headers
+//  - keyName: string representing the keyName, see https://docs.microsoft.com/en-us/rest/api/keyvault/encrypt/encrypt#uri-parameters
+//  - keyVersion: string representing the keyVersion, see https://docs.microsoft.com/en-us/rest/api/keyvault/encrypt/encrypt#uri-parameters
+//  - opts: *KeeperOptions with the desired Algorithm to use for operations. See this link for more info: https://docs.microsoft.com/en-us/rest/api/keyvault/encrypt/encrypt#jsonwebkeyencryptionalgorithm
+func NewKeeper(client *keyvault.BaseClient, keyVaultName, keyName, keyVersion string, opts *KeeperOptions) *secrets.Keeper {
+	return secrets.NewKeeper(&keeper{
+		client:       client,
+		keyVaultName: keyVaultName,
+		keyName:      keyName,
+		keyVersion:   keyVersion,
+		options:      opts,
+	})
+}
+
+// Encrypt encrypts the plaintext into a ciphertext.
+func (k *keeper) Encrypt(ctx context.Context, plaintext []byte) ([]byte, error) {
+	if err := k.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	b64Text := base64.StdEncoding.EncodeToString(plaintext)
+	keyOpsResult, err := k.client.Encrypt(ctx, k.getKeyVaultURI(), k.keyName, k.keyVersion, keyvault.KeyOperationsParameters{
+		Algorithm: keyvault.JSONWebKeyEncryptionAlgorithm(k.options.Algorithm),
```

Comment:
`options` could be `nil`, I think we should give a default algorithm if user doesn't want to provide one.

Reply:
In this case it shouldn't, the user should choose the appropriate algorithm. Omitting / nil option will display a message listing all acceptable algorithms.

Reply:
That's OK, and we should move call `validateOptions` here since it applies to both `Encrypt` and `Decrypt`.